REFACTORING_RECOMMENDATIONS.md

```diff
@@ -0,0 +1,1205 @@
+**Lines Saved**: ~29 (50% reduction)
+**Effort**: 6-8 hours
+
+**Implementation**:
+
+```typescript
+// Modify base Session class: /packages/protocol/src/session/Session.ts
+export abstract class Session<T extends Session.Events = Session.Events> {
+    protected isDestroyed = false;
+
+    /**
+     * Template method for session cleanup.
+     * Subclasses can override hooks for custom behavior.
+     */
+    protected end(): void {
+        if (this.isDestroyed) {
+            return;
+        }
+
+        this.isDestroyed = true;
+
+        // Hook for pre-cleanup actions
+        this.beforeEnd();
+
+        // Standard cleanup
+        this.context.close();
+
+        // Hook for post-cleanup actions
+        this.afterEnd();
+    }
+
+    /**
+     * Override to perform actions before context close.
+     */
+    protected beforeEnd(): void {}
+
+    /**
+     * Override to perform actions after context close.
+     */
+    protected afterEnd(): void {}
+}
+```
+
+**Refactor Implementations**:
+
+```typescript
+// Before: /packages/protocol/src/session/InsecureSession.ts
+end() {
+    if (this.#isDestroyed) return;
+    this.#isDestroyed = true;
+    this.context.close();
+    this.session?.end();
+}
+
+// After:
+protected afterEnd(): void {
+    this.session?.end();
+}
+
+// The base end() method handles isDestroyed check and context.close()
+```
+
+```typescript
+// Before: /packages/protocol/src/session/NodeSession.ts
+async destroy(shouldSendCloseSession = false): Promise<void> {
+    if (this.#isDestroyed) return;
+    this.#isDestroyed = true;
+    await this.#closingAfterExchangeFinished;
+    if (shouldSendCloseSession) {
+        await this.sendCloseSession();
+    }
+    this.context.close();
+    this.session?.end();
+}
+
+// After:
+async destroy(shouldSendCloseSession = false): Promise<void> {
+    if (this.isDestroyed) return;
+
+    this.isDestroyed = true;
+    await this.#closingAfterExchangeFinished;
+
+    if (shouldSendCloseSession) {
+        await this.sendCloseSession();
+    }
+
+    this.end();
+}
+
+protected afterEnd(): void {
+    this.session?.end();
+}
+```
+
+**Testing**:
+- Unit tests for session lifecycle
+- Test destroy with/without close session
+- Test end() is idempotent
+- Integration tests for session cleanup
+- Verify no resource leaks
+
+**Files to Refactor**:
+1. `session/NodeSession.ts:328-367`
+2. `session/InsecureSession.ts:92-100`
+3. `session/GroupSession.ts:218-225`
+
+---
+
+### 2.4 Create SessionCrypto Utility
+
+**Files Affected**: 2 session files
+**Lines Saved**: ~23 (74% reduction)
+**Effort**: 4-6 hours
+
+**Implementation**:
+
+```typescript
+// File: /packages/protocol/src/session/SessionCrypto.ts
+import { Crypto } from "#crypto";
+import { Bytes } from "#general";
+
+/**
+ * Utility class for session encryption/decryption operations.
+ * Centralizes nonce generation and crypto operations used across session types.
+ */
+export class SessionCrypto {
```

Comment:
These should probably just be moved into SecureSession instead?  It's already common base class...